Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
-      }
-      lboard.entries.insert_or_assign(org_entry.index, std::move(dest_entry));
-      if (org_entry.username == username)
-        lboard.player_index = org_entry.index;
-    }
-    rc_api_destroy_fetch_leaderboard_info_response(&board_info);
-  }
+void AchievementManager::HandleLeaderboardStartedEvent(const rc_client_event_t* client_event)
+{
+  OSD::AddMessage(fmt::format("Attempting leaderboard: {} - {}", client_event->leaderboard->title,
+                              client_event->leaderboard->description),
+                  OSD::Duration::VERY_LONG, OSD::Color::GREEN);
+  AchievementManager::GetInstance().FetchBoardInfo(client_event->leaderboard->id);
+}
 
-  {
-    std::lock_guard lg{m_lock};
-    m_leaderboard_map.insert_or_assign(leaderboard_id, std::move(lboard));
-  }
+void AchievementManager::HandleLeaderboardFailedEvent(const rc_client_event_t* client_event)
+{
+  OSD::AddMessage(fmt::format("Failed leaderboard: {}", client_event->leaderboard->title),
+                  OSD::Duration::VERY_LONG, OSD::Color::RED);
+  AchievementManager::GetInstance().FetchBoardInfo(client_event->leaderboard->id);
+}
 
-  return ResponseType::SUCCESS;
+void AchievementManager::HandleLeaderboardSubmittedEvent(const rc_client_event_t* client_event)
+{
+  OSD::AddMessage(fmt::format("Scored {} on leaderboard: {}",
+                              client_event->leaderboard->tracker_value,
+                              client_event->leaderboard->title),
+                  OSD::Duration::VERY_LONG, OSD::Color::YELLOW);
+  AchievementManager::GetInstance().FetchBoardInfo(client_event->leaderboard->id);
 }
 
-void AchievementManager::ActivateDeactivateAchievement(AchievementId id, bool enabled,
-                                                       bool unofficial, bool encore)
+void AchievementManager::HandleLeaderboardTrackerUpdateEvent(const rc_client_event_t* client_event)
 {
-  auto it = m_unlock_map.find(id);
-  if (it == m_unlock_map.end())
-  {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Attempted to unlock unknown achievement id {}.", id);
-    return;
-  }
-  const UnlockStatus& status = it->second;
-  u32 index = status.game_data_index;
-  bool active = (rc_runtime_get_achievement(&m_runtime, id) != nullptr);
-  bool hardcore_mode_enabled = Config::Get(Config::RA_HARDCORE_ENABLED);
-
-  // Deactivate achievements if game is not loaded
-  bool activate = m_is_game_loaded;
-  // Activate achievements only if achievements are enabled
-  if (activate && !enabled)
-    activate = false;
-  // Deactivate if achievement is unofficial, unless unofficial achievements are enabled
-  if (activate && !unofficial &&
-      m_game_data.achievements[index].category == RC_ACHIEVEMENT_CATEGORY_UNOFFICIAL)
+  auto& active_leaderboards = AchievementManager::GetInstance().m_active_leaderboards;
+  for (auto itr = active_leaderboards.begin(); itr < active_leaderboards.end(); itr++)
```

Comment:
Any reason we aren't doing:

`for (const auto& leaderboard : active_leaderboards)`

?

Reply:
Old habits die hard.